src/psyclone/psyGen.py

```diff
@@ -1914,6 +1928,49 @@ def _encloses_omp_directive(self):
             #                       "any OpenMP directives. This is probably "
             #                       "not what you want.")
 
+    def update(self):
+        ''' 
+        Updates the fparser2 AST by inserting nodes for this OpenMP
+        parallel region.
+
+        :raises InternalError: if the existing AST doesn't have the \
+                               correct structure to permit the insertion \
+                               of the OpenMP parallel region.
+        '''
+        from fparser.common.readfortran import FortranStringReader
+        from fparser.two.Fortran2003 import Comment
+        # Check that we haven't already been called
+        if self._ast:
+            return
+        # Find the locations in which we must insert the begin/end
+        # directives...
+        # Find the children of this node in the AST of our parent node
+        try:
```

Comment:
I'm sure you don't need me to tell you that this insertion is a bit messy and we need to better support this sort of thing in fparser2, especially adding directives.

Reply:
Agreed.